### ISSUE TITLE
SSL Engine is not available with BoringSSL

### ISSUE THREAD

**Issue by maskit** [apache/trafficserver#7512]
BoringSSL doesn't have Engine API. Autest tls_engine fails because of this.

Engine API is going to be deprecated since OpenSSL 3.0, and Provider seems to be a replacement.

**Comment by alokrahul513** [comment]
Hi, I want to work on this issue. can you help me to get started?

**Comment by bneradt** [comment]
When working through test and build updates needed to support fedora:41, I ran into this Engine deprecation issue when building against fedora:41's system openssl:

```
[356/1286] Building CXX object src/iocore/net/CMakeFiles/inknet.dir/SSLUtils.cc.o
FAILED: src/iocore/net/CMakeFiles/inknet.dir/SSLUtils.cc.o 
/usr/bin/ccache /usr/lib64/ccache/c++ -DDEBUG -DOPENSSL_API_COMPAT=10002 -DOPENSSL_IS_OPENSSL3 -DPACKAGE_NAME="\"Apache Traffic Server\"" -DPACKAGE_VERSION=\"10.1.0\" -D_DEBUG -Dlinux -I/home/jenkins/trafficserver/include -I/home/jenkins/trafficserver/build/include -I/home/jenkins/trafficserver/lib/swoc/include -I/home/jenkins/trafficserver/lib/yamlcpp/include -isystem /home/jenkins/trafficserver/lib/systemtap -pthread -g -std=c++20 -Wno-invalid-offsetof -pipe -Wall -Wextra -Wno-noexcept-type -Wsuggest-override -Wno-vla-extension -fno-strict-aliasing -Wno-format-truncation -Wno-deprecated-declarations -mcx16 -Werror -MD -MT src/iocore/net/CMakeFiles/inknet.dir/SSLUtils.cc.o -MF src/iocore/net/CMakeFiles/inknet.dir/SSLUtils.cc.o.d -o src/iocore/net/CMakeFiles/inknet.dir/SSLUtils.cc.o -c /home/jenkins/trafficserver/src/iocore/net/SSLUtils.cc
/home/jenkins/trafficserver/src/iocore/net/SSLUtils.cc:69:10: fatal error: openssl/engine.h: No such file or directory
   69 | #include <openssl/engine.h>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
[361/1286] Building CXX object src/iocore/net/CMakeFiles/inknet.dir/OCSPStapling.cc.o
ninja: build stopped: subcommand failed.
```

This fedora change is described here:
https://discussion.fedoraproject.org/t/f41-change-proposal-openssl-deprecate-engine-system-wide/111344

The `openssl` and `openssl-devel` packages no longer provide engine.h as a part of fedora:41. They removed it because the engine.h API has been deprecated since the beginning of OpenSSL 3.0. For the time being to unblock updating CI to fedora:41, I'll install the `openssl-devel-engine` package for fedora:41. But it would be good for us to transition off of the deprecated Engine API and move toward the Provider API per @maskit's suggestion above.

**Comment by jeredfloyd** [comment]
This will also be an issue for RHEL 10, which will not include an openssl-devel-engine package like Fedora 41: https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/10-beta/html-single/10.0_beta_release_notes/index#deprecated-features-security

**Comment by bryancall** [comment]
I'm closing this as resolved. The concrete problems here, the Engine-based AuTest failing and the build breaking where the OpenSSL Engine API is unavailable (BoringSSL, Fedora 41, and RHEL 10), are fixed. PR https://github.com/apache/trafficserver/pull/13264 (first released in 10.1.3) removed the dynamic Engine dependency from the test coverage: it deleted the old async_engine test plugin and replaced it with a plugin that exercises the SSL_ERROR_WANT_ASYNC handshake path without the Engine API. The remaining runtime Engine calls in SSLUtils.cc are now guarded behind HAVE_ENGINE_* build feature checks, so the source compiles cleanly on builds with no Engine support. A complete migration of runtime key loading to the OpenSSL Provider API is a separate enhancement and is no longer needed to unblock these builds, so it is better tracked as its own issue if anyone wants to pursue it. Please reopen if a build or test still fails on an Engine-less OpenSSL.
